In [83]:
import pandas as pd
import numpy as np
import time
from collections import Counter

pd.set_option('display.max_rows', 10000)
pd.set_option('display.max_columns', 1000)

In [76]:
data = pd.read_csv('/content/drive/MyDrive/CSE572 Data Mining/HW3/data.csv', header=None)
labels = pd.read_csv('/content/drive/MyDrive/CSE572 Data Mining/HW3/label.csv', header=None)

In [77]:
data.head()

0    1    2    3    4    5    6    7    8    9    10   11   12   13   14   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   15   16   17   18   19   20   21   22   23   24   25   26   27   28   29   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   30   31   32   33   34   35   36   37   38   39   40   41   42   43   44   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   45   46   47   48   49   50   51   52   53   54   55   56   57   58   59   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   60   61   62   63   64   65   66   67   68   69   70   71   72   73   74   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   75   76   77   78   79   80   81   82   83   84   85   86   87   88   89   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   90   91   92   93   94   95   96   97   98   99   100  101  102  103  104  \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0  116  125  171  255  255  150   93    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   105  106  107  108  109  110  111  112  113  114  115  116  117  118  119  \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   120  121  122  123  124  125  126  127  128  129  130  131  132  133  134  \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    

In [4]:
data.shape

(10000, 784)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Columns: 784 entries, 0 to 783
dtypes: int64(784)
memory usage: 59.8 MB


In [6]:
data.describe()

0        1        2        3        4        5        6        7    \
count  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
mean       0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
std        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
min        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
25%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
50%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
75%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
max        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

           8        9        10       11       12       13       14       15   \
count  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
mean       0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
std        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
min        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
25%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
50%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
75%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
max        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

           16       17       18       19       20       21       22       23   \
count  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
mean       0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
std        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
min        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
25%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
50%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
75%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
max        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

           24       25       26       27       28       29       30       31   \
count  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0  10000.0   
mean       0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
std        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
min        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
25%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
50%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
75%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
max        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

           32          33          34          35            36   \
count  10000.0  10000.0000  10000.0000  10000.0000  10000.000000   
mean       0.0      0.0038      0.0236      0.0158      0.040300   
std        0.0      0.3800      2.3600      1.5800      2.884113   
min        0.0      0.0000      0.0000      0.0000      0.000000   
25%        0.0      0.0000      0.0000      0.0000      0.000000   
50%        0.0      0.0000      0.0000      0.0000      0.000000   
75%        0.0      0.0000      0.0000      0.0000      0.000000   
max        0.0     38.0000    236.0000    158.0000    233.000000   

                37            38            39            40           41   \
count  10000.000000  10000.000000  10000.000000  10000.000000  10000.00000   
mean       0.051100      0.104700      0.234300      0.308800      0.26600   
std        2.699336      4.516213      6.495508      7.865522      6.99538   
min        0.000000      0.000000      0.000000      0.000000      0.00000   
25%        0.000000      0.000000      0.000000      0.000000      0.00000   
50%       

In [93]:
data = pd.read_csv('/content/drive/MyDrive/CSE572 Data Mining/HW3/data.csv', header=None).to_numpy()
labels = pd.read_csv('/content/drive/MyDrive/CSE572 Data Mining/HW3/label.csv', header=None).to_numpy().flatten()

**Q1: Run K-means clustering with Euclidean, Cosine and Jarcard similarity. Specify K= the number of categorical values of y (the number of classifications). Compare the SSEs of Euclidean-K-means, Cosine-K-means,Jarcard-K-means. Which method is better?  (10 points)**

In [79]:
# Number of clusters based on labels
k = len(np.unique(labels))
print(f'Count of unique labels in label.csv aka number of clusters is {k}')

Count of unique labels in label.csv aka number of clusters is 10


In [80]:
# Distance metrics calculations
def euclidean_distance(x, y):
    return np.linalg.norm(x - y)

def cosine_distance(x, y):
    return 1 - np.dot(x, y) / (np.linalg.norm(x) * np.linalg.norm(y))

def jaccard_distance(x, y):
    numerator = np.minimum(x, y).sum()
    denominator = np.maximum(x, y).sum()
    return 1 - (numerator / denominator) if denominator != 0 else 1

In [81]:
# Calculate Sum of Squared Errors
def compute_sse(X, clusters, centroids):
    sse = 0
    for cluster_idx, cluster in enumerate(clusters):
        for point_idx in cluster:
            sse += euclidean_distance(X[point_idx], centroids[cluster_idx]) ** 2
    return sse

In [82]:
# Custom K-Means algorithm
def kmeans_clustering(X, k, distance_fn, max_iters=500, tol=1e-4):
    np.random.seed(42)
    centroids = X[np.random.choice(X.shape[0], k, replace=False)]  # Initialize centroids

    for iteration in range(max_iters):
        # Step 1: Assign points to the nearest centroid
        clusters = [[] for _ in range(k)]
        for idx, point in enumerate(X):
            distances = [distance_fn(point, centroid) for centroid in centroids]
            cluster_idx = np.argmin(distances)
            clusters[cluster_idx].append(idx)

        # Step 2: Update centroids
        new_centroids = []
        for cluster in clusters:
            if cluster:  # Avoid empty clusters
                new_centroids.append(np.mean(X[cluster], axis=0))
            else:
                new_centroids.append(centroids[np.random.choice(range(k))])  # Reinitialize empty clusters
        new_centroids = np.array(new_centroids)

        # Step 3: Check for convergence
        centroid_shift = np.sum([euclidean_distance(old, new) for old, new in zip(centroids, new_centroids)])
        if centroid_shift < tol:
            break
        centroids = new_centroids

    return clusters, centroids

In [48]:
def run_experiment(X, labels, k):
    results = {}
    distance_metrics = {
        "Euclidean Distance": euclidean_distance,
        "Cosine Similarity": cosine_distance,
        "Jaccard Similarity": jaccard_distance,
    }

    for name, distance_fn in distance_metrics.items():
        clusters, centroids = kmeans_clustering(X, k, distance_fn)
        sse = compute_sse(X, clusters, centroids)
        results[name] = {"SSE": sse}

    return results

In [49]:
# Run the experiment
results = run_experiment(data, labels, k)

# Print results
for metric, stats in results.items():
    print(f"{metric} - SSE: {stats['SSE']:.2f}")

Euclidean Distance - SSE: 25414767689.96
Cosine Similarity - SSE: 25489564392.28
Jaccard Similarity - SSE: 25416346592.39


SSE of Euclidean Distance = 25414767689.96

SSE of Cosine Similarity = 25489564392.28

SSE of Jaccard Similarity = 25416346592.39

From the above results, we can infer that **Euclidean** Distance is the best method as it provides the lowest SSE value.

**Q2: Compare the accuracies of Euclidean-K-means Cosine-K-means, Jarcard-K-means. First, label each cluster using the majority vote label of the data points in that cluster. Later, compute the predictive accuracy of Euclidean-K-means, Cosine-K-means, Jarcard-K-means. Which metric is better? (10 points)**

In [55]:
# Step 1 - Label each cluster using majority vote
def majority_vote(labels, clusters):
    cluster_labels = []
    for cluster in clusters:
        if cluster:
            cluster_label = Counter(labels[cluster]).most_common(1)[0][0]
        else:
            cluster_label = -1  # Handle empty clusters
        cluster_labels.append(cluster_label)
    return cluster_labels

In [56]:
# Step 2 - Compute accuracy of each metric
def compute_accuracy(labels, clusters, cluster_labels):
    correct = 0
    total = 0
    for cluster_idx, cluster in enumerate(clusters):
        for point_idx in cluster:
            if labels[point_idx] == cluster_labels[cluster_idx]:
                correct += 1
            total += 1
    return correct / total

In [57]:
def run_experiment(X, labels, k):
    results = {}
    distance_metrics = {
        "Euclidean": euclidean_distance,
        "Cosine": cosine_distance,
        "Jaccard": jaccard_distance,
    }

    for name, distance_fn in distance_metrics.items():
        clusters, centroids = kmeans_clustering(X, k, distance_fn)
        cluster_labels = majority_vote(labels, clusters)
        accuracy = compute_accuracy(labels, clusters, cluster_labels)
        results[name] = {"Accuracy": accuracy}

    return results

In [61]:
results = run_experiment(data, labels, k)

# Print results
for metric, stats in results.items():
    print(f"{metric} - Accuracy: {stats['Accuracy']:.2%}")

Euclidean - Accuracy: 58.51%
Cosine - Accuracy: 63.09%
Jaccard - Accuracy: 60.21%


Accuracy of Euclidean Distance is 58.51 %

Accuracy of Cosine Similarity is 63.09 %

Accuracy of Jaccard Similarity is 60.21 %

From the above results, we can infer that **Cosine Similarity is the best metric** as it provides the highest accuracy.

**Q3: Set up the same stop criteria: “when there is no change in centroid position OR when the SSE value increases in the next iteration OR when the maximum preset value (e.g., 500, you can set the preset value by yourself) of iteration is complete”,  for Euclidean-K-means, Cosine-K means, Jarcard-K-means. Which method requires more iterations and times to converge? (10 points)**

In [85]:
# Modified K-means with new stopping criteria
def k_means_with_criteria(data, k, distance_function, max_iters=100):
    n_samples = data.shape[0]
    centroids = data[np.random.choice(n_samples, k, replace=False)]
    prev_centroids = np.zeros_like(centroids)
    labels = np.zeros(n_samples, dtype=int)
    prev_sse = float("inf")
    iteration = 0

    start_time = time.time()

    for iteration in range(1, max_iters + 1):
        # Step 1: Assign labels based on closest centroid
        distances = np.array([[distance_function(sample, centroid) for centroid in centroids] for sample in data])
        labels = np.argmin(distances, axis=1)

        # Step 2: Recompute centroids
        new_centroids = np.array([data[labels == j].mean(axis=0) if np.any(labels == j) else centroids[j] for j in range(k)])

        # Step 3: Compute SSE
        sse = sum(np.sum((data[labels == j] - centroids[j]) ** 2) for j in range(k))

        # Step 4: Check stopping criteria
        if np.allclose(new_centroids, centroids, atol=1e-6):
            print(f"Converged: No change in centroids at iteration {iteration}")
            break
        if sse > prev_sse:
            print(f"Stopped: SSE increased at iteration {iteration}")
            break

        centroids = new_centroids
        prev_sse = sse

    end_time = time.time()

    # Return results
    return {
        "clusters": [data[labels == j] for j in range(k)],
        "centroids": centroids,
        "sse": sse,
        "iterations": iteration,
        "time": end_time - start_time
    }

# Run K-means for each distance metric and compare iterations and time
results = {}
for metric_name, distance_func in {
    'Euclidean': euclidean_distance,
    'Cosine': cosine_distance,
    'Jaccard': jaccard_distance,
}.items():
    print(f"Running K-means with {metric_name} distance...")
    result = k_means_with_criteria(data, k, distance_func)
    results[metric_name] = result
    print(f"{metric_name}: Iterations = {result['iterations']}, Time = {result['time']:.2f} seconds")

# Summary of results
print("\nSummary of Results:")
for metric_name, result in results.items():
    print(f"{metric_name}: Iterations = {result['iterations']}, Time = {result['time']:.2f} seconds")

Running K-means with Euclidean distance...
Converged: No change in centroids at iteration 75
Euclidean: Iterations = 75, Time = 83.05 seconds
Running K-means with Cosine distance...
Stopped: SSE increased at iteration 31
Cosine: Iterations = 31, Time = 53.52 seconds
Running K-means with Jaccard distance...
Stopped: SSE increased at iteration 27
Jaccard: Iterations = 27, Time = 48.33 seconds

Summary of Results:
Euclidean: Iterations = 75, Time = 83.05 seconds
Cosine: Iterations = 31, Time = 53.52 seconds
Jaccard: Iterations = 27, Time = 48.33 seconds


**Euclidean K-Means** requires the most iterations and the most time to converge.

In [89]:
# Modified K-means to support multiple stopping conditions
def k_means_multiple_stops(data, k, distance_function, max_iters=100):
    n_samples, n_features = data.shape
    centroids = data[np.random.choice(n_samples, k, replace=False)]
    prev_centroids = np.zeros_like(centroids)
    labels = np.zeros(n_samples, dtype=int)
    prev_sse = float("inf")
    sse_stopping_conditions = {}

    start_time = time.time()

    for iteration in range(1, max_iters + 1):
        # Step 1: Assign labels based on closest centroid
        distances = np.linalg.norm(data[:, np.newaxis] - centroids, axis=2)
        labels = np.argmin(distances, axis=1)

        # Step 2: Recompute centroids
        new_centroids = np.array([
            data[labels == j].mean(axis=0) if np.any(labels == j) else centroids[j]
            for j in range(k)
        ])

        # Step 3: Compute SSE
        sse = sum(np.sum((data[labels == j] - centroids[j]) ** 2) for j in range(k))

        # Step 4: Check each stopping condition
        if "centroid_no_change" not in sse_stopping_conditions and np.allclose(new_centroids, centroids, atol=1e-6):
            sse_stopping_conditions["centroid_no_change"] = (iteration, sse)

        if "sse_increase" not in sse_stopping_conditions and sse > prev_sse:
            sse_stopping_conditions["sse_increase"] = (iteration, prev_sse)
            break  # SSE increase ends the run

        centroids = new_centroids
        prev_sse = sse

    # Stopping due to max iterations
    if "max_iterations" not in sse_stopping_conditions:
        sse_stopping_conditions["max_iterations"] = (iteration, sse)

    end_time = time.time()

    return {
        "sse_conditions": sse_stopping_conditions,
        "time": end_time - start_time
    }

# Run K-means for each distance metric and stopping condition
results = {}
for metric_name, distance_func in {
    'Euclidean': euclidean_distance,
    'Cosine': cosine_distance,
    'Jaccard': jaccard_distance,
}.items():
    print(f"Running K-means with {metric_name} distance...")
    result = k_means_multiple_stops(data, k, distance_func)
    results[metric_name] = result
    print(f"{metric_name}: {result['sse_conditions']}")

# Output SSE comparisons
print("\nSSE Comparisons for Different Stopping Conditions:")
for metric_name, result in results.items():
    print(f"\n{metric_name}:")
    print(f"  Time taken: {result['time']:.2f} seconds")
    for condition, (iterations, sse) in result["sse_conditions"].items():
        print(f"  {condition}: Iterations = {iterations}, SSE = {sse:.2f}")

Running K-means with Euclidean distance...
Euclidean: {'centroid_no_change': (35, 25454826721.149597), 'max_iterations': (100, 25454826721.149597)}
Running K-means with Cosine distance...
Cosine: {'centroid_no_change': (86, 25320586006.17389), 'max_iterations': (100, 25320586006.17389)}
Running K-means with Jaccard distance...
Jaccard: {'centroid_no_change': (76, 25321599262.60322), 'max_iterations': (100, 25321599262.60322)}

SSE Comparisons for Different Stopping Conditions:

Euclidean:
  Time taken: 82.96 seconds
  centroid_no_change: Iterations = 35, SSE = 25454826721.15
  max_iterations: Iterations = 100, SSE = 25454826721.15

Cosine:
  Time taken: 82.76 seconds
  centroid_no_change: Iterations = 86, SSE = 25320586006.17
  max_iterations: Iterations = 100, SSE = 25320586006.17

Jaccard:
  Time taken: 84.45 seconds
  centroid_no_change: Iterations = 76, SSE = 25321599262.60
  max_iterations: Iterations = 100, SSE = 25321599262.60


In [96]:
# Modified K-means to support multiple stopping conditions
def k_means_multiple_stops(data, k, distance_function, max_iters=100):
    n_samples, n_features = data.shape
    centroids = data[np.random.choice(n_samples, k, replace=False)]
    prev_centroids = np.zeros_like(centroids)
    labels = np.zeros(n_samples, dtype=int)
    prev_sse = float("inf")
    sse_stopping_conditions = {}

    start_time = time.time()

    for iteration in range(1, max_iters + 1):
        # Step 1: Assign labels based on closest centroid
        distances = distance_function(data, centroids)  # (n_samples, k)
        labels = np.argmin(distances, axis=1)  # Assign labels based on the closest centroid

        # Step 2: Recompute centroids
        new_centroids = np.array([
            data[labels == j].mean(axis=0) if np.any(labels == j) else centroids[j]
            for j in range(k)
        ])

        # Step 3: Compute SSE
        sse = sum(np.sum((data[labels == j] - centroids[j]) ** 2) for j in range(k))

        # Step 4: Check each stopping condition
        if "centroid_no_change" not in sse_stopping_conditions and np.allclose(new_centroids, centroids, atol=1e-6):
            sse_stopping_conditions["centroid_no_change"] = (iteration, sse)

        if "sse_increase" not in sse_stopping_conditions and sse > prev_sse:
            sse_stopping_conditions["sse_increase"] = (iteration, prev_sse)

        centroids = new_centroids
        prev_sse = sse

    # Stopping due to max iterations
    if "max_iterations" not in sse_stopping_conditions:
        sse_stopping_conditions["max_iterations"] = (iteration, sse)

    end_time = time.time()

    return {
        "sse_conditions": sse_stopping_conditions,
        "time": end_time - start_time
    }

# Run K-means for each distance metric and stopping condition
results = {}
for metric_name, distance_func in {
    'Euclidean': euclidean_distance,
    'Cosine': cosine_distance,
    'Jaccard': jaccard_distance,
}.items():
    print(f"Running K-means with {metric_name} distance...")
    result = k_means_multiple_stops(data, k, distance_func)
    results[metric_name] = result
    print(f"{metric_name}: {result['sse_conditions']}")

# Output SSE comparisons
print("\nSSE Comparisons for Different Stopping Conditions:")
for metric_name, result in results.items():
    print(f"\n{metric_name}:")
    print(f"  Time taken: {result['time']:.2f} seconds")
    for condition, (iterations, sse) in result["sse_conditions"].items():
        print(f"  {condition}: Iterations = {iterations}, SSE = {sse:.2f}")


Running K-means with Euclidean distance...
Euclidean: {'centroid_no_change': (55, 25418063136.679237), 'max_iterations': (100, 25418063136.679237)}
Running K-means with Cosine distance...
Cosine: {'sse_increase': (37, 25461655725.744656), 'centroid_no_change': (95, 25419129705.29959), 'max_iterations': (100, 25419129705.29959)}
Running K-means with Jaccard distance...
Jaccard: {'sse_increase': (15, 25628860516.507458), 'centroid_no_change': (74, 25416922861.342865), 'max_iterations': (100, 25416922861.342865)}

SSE Comparisons for Different Stopping Conditions:

Euclidean:
  Time taken: 86.12 seconds
  centroid_no_change: Iterations = 55, SSE = 25418063136.68
  max_iterations: Iterations = 100, SSE = 25418063136.68

Cosine:
  Time taken: 31.01 seconds
  sse_increase: Iterations = 37, SSE = 25461655725.74
  centroid_no_change: Iterations = 95, SSE = 25419129705.30
  max_iterations: Iterations = 100, SSE = 25419129705.30

Jaccard:
  Time taken: 100.44 seconds
  sse_increase: Iterations =

Euclidean Distance:

SSE for "when there is no change in centroid position" criterion = 25418063136.68

SSE for "when the max present value of iteration is complete" criterion = 25418063136.68
<br>

Cosine Similarity:

SSE for "when there is no change in centroid position" criterion = 25419129705.30

SSE for "when the SSE value increases in the next iteration" criterion = 25461655725.74

SSE for "when the max present value of iteration is complete" criterion = 25419129705.30
<br>

Jaccard Similarity:

SSE for "when there is no change in centroid position" criterion = 25416922861.34

SSE for "when the SSE value increases in the next iteration" criterion = 25628860516.51

SSE for "when the max present value of iteration is complete" criterion = 25416922861.34

**Q5: What are your summary observations or takeaways based on your algorithmic analysis? (5 points)**

Euclidean K-Means method is best for achieving the lowest SSE but takes the most iterations and time because calculation of Euclidean distance is more computationally expensive for high dimensional data like the one provided in data.csv in kmeans_data.zip

Cosine K-Means method gives the best balance and trade-off between time, accuracy and SSE for high dimensional data.

Euclidean K-Means method is never satisfies "when the SSE value increases in the next iteration" criterion because by default, the nature of Euclidean K-Means will never result in an increase in SSE.